udata/harvest/backends/dcat.py

```diff
@@ -154,11 +157,35 @@ def walk_graph(self, url: str, fmt: str) -> Generator[tuple[int, Graph], None, N
     def process_one_datasets_page(self, page_number: int, page: Graph):
         for node in page.subjects(RDF.type, DCAT.Dataset):
             remote_id = page.value(node, DCT.identifier)
+            if self.is_dataset_inside_dataservice_graph(page, node):
+                continue
+
             self.process_dataset(remote_id, page_number=page_number, page=page, node=node)
 
             if self.is_done():
                 return
 
+    def is_dataset_inside_dataservice_graph(self, page: Graph, dataset_node) -> bool:
+        # In dataservice nodes we have `servesDataset` or `hasPart` that can contains nodes
+        # with type=dataset. We don't want to process them because these nodes are empty (they
+        # only contains a link to the dataset definition).
+        # These datasets are either present in the catalog in past or future pages or datasets
+        # external from the catalog we are currently harvesting (so we don't want to harvest them).
+        # First we thought of skipping them inside `dataset_from_rdf` (see :ExcludeExternalyDefinedDataset)
+        # but it creates a lot of "fake" items in the job and raising problems (reaching the max harvest item for
+        # example and not getting to the "real" datasets/dataservices in subsequent pages)
+        # We could check if the node is really empty before discarding it?
+        for node in page.subjects(RDF.type, DCAT.DataService):
+            dataservice = page.resource(node)
+            for dataset in dataservice.objects(DCAT.servesDataset):
+                if dataset_node == dataset.identifier:
```

Comment:
Are we sure it works if a dataset is properly described in the catalog but is also served by a dataservice?

Reply:
No it doesn't work in some case (I was waiting for the tests and they fail…), need to find a better way to exclude them…